test/002_highsnr_ping_pong/ping.py

```diff
@@ -0,0 +1,199 @@
+AUDIO_INPUT_DEVICE = args.AUDIO_INPUT
+
+# 1024 good for mode 6
+AUDIO_FRAMES_PER_BUFFER = 2048 
+MODEM_SAMPLE_RATE = 8000
+
+FREEDV_TX_MODE = args.FREEDV_TX_MODE
+FREEDV_RX_MODE = args.FREEDV_RX_MODE
+
+DEBUGGING_MODE = args.DEBUGGING_MODE
+        #-------------------------------------------- LOAD FREEDV        
+libname = pathlib.Path().absolute() / "codec2/build_linux/src/libcodec2.so"
+c_lib = ctypes.CDLL(libname)
+
+        #--------------------------------------------CREATE PYAUDIO  INSTANCE        
+p = pyaudio.PyAudio()
+        #--------------------------------------------GET SUPPORTED SAMPLE RATES FROM SOUND DEVICE        
+#AUDIO_SAMPLE_RATE_TX = int(p.get_device_info_by_index(AUDIO_OUTPUT_DEVICE)['defaultSampleRate'])
+#AUDIO_SAMPLE_RATE_RX = int(p.get_device_info_by_index(AUDIO_INPUT_DEVICE)['defaultSampleRate']) 
+AUDIO_SAMPLE_RATE_TX = 8000
+AUDIO_SAMPLE_RATE_RX = 8000       
+        #--------------------------------------------OPEN AUDIO CHANNEL TX
+
+stream_tx = p.open(format=pyaudio.paInt16,
+                            channels=1,
+                            rate=AUDIO_SAMPLE_RATE_TX,
+                            frames_per_buffer=AUDIO_FRAMES_PER_BUFFER, #n_nom_modem_samples
+                            output=True,
+                            output_device_index=AUDIO_OUTPUT_DEVICE,  
+                            )  
+             
+stream_rx = p.open(format=pyaudio.paInt16, 
+                            channels=1,
+                            rate=AUDIO_SAMPLE_RATE_RX,
+                            frames_per_buffer=AUDIO_FRAMES_PER_BUFFER,
+                            input=True,
+                            input_device_index=AUDIO_INPUT_DEVICE,
+                            ) 
+    
+       
+                            
+
+                         
+def receive():
+
+    c_lib.freedv_open.restype = ctypes.POINTER(ctypes.c_ubyte)
+    freedv = c_lib.freedv_open(FREEDV_RX_MODE)
+    bytes_per_frame = int(c_lib.freedv_get_bits_per_modem_frame(freedv)/8)
+    payload_per_frame = bytes_per_frame -2
+    n_nom_modem_samples = c_lib.freedv_get_n_nom_modem_samples(freedv)
+    n_tx_modem_samples = c_lib.freedv_get_n_tx_modem_samples(freedv) #get n_tx_modem_samples which defines the size of the modulation object # --> *2
+
+    bytes_out = (ctypes.c_ubyte * bytes_per_frame) #bytes_per_frame
+    bytes_out = bytes_out() #get pointer from bytes_out  
+    
+    total_n_bytes = 0
+    rx_total_frames = 0
+    rx_frames = 0
+    rx_bursts = 0
+    receive = True
+    while receive == True:
+        time.sleep(0.01)
+   
+        nin = c_lib.freedv_nin(freedv)
+        nin_converted = int(nin*(AUDIO_SAMPLE_RATE_RX/MODEM_SAMPLE_RATE))
+        if DEBUGGING_MODE == True:
+            print("-----------------------------")
+            print("NIN:  " + str(nin) + " [ " + str(nin_converted) + " ]")
+        
+        data_in = stream_rx.read(nin_converted,  exception_on_overflow = False)  
+        data_in = data_in.rstrip(b'\x00')
```

Comment:
I'd suggest a design that [reads a constant buffer size](https://github.com/drowe67/codec2/pull/265/files#diff-194711ca5dcb53842b73b19af641619f1065e2390e8b302218adaa9e85aff4e5) from the audio card, then uses that to fill a buffer up to `nin` samples.  That's just a gut feel, based on experience I guess (and mistrust of Linux audio!).

Reply:
I think I already did this with my multimode test :thinking: . I need to change it so it fits to our tests. I will do this the next days.

Reply:
I will also try doing this within our single mode tests. We are also working out a good example/demo for learning codec2 by doing this.
I'll add this to the todo list.